app/uk/gov/hmrc/transitmovements/repositories/MovementsRepository.scala

```diff
@@ -548,4 +543,46 @@ class MovementsRepositoryImpl @Inject() (
         Future.successful(Left(UnexpectedError(Some(ex))))
     })
   }
+
+  def isMessageIdUnique(messageId: MessageId): EitherT[Future, MongoError, Unit] = {
+    val selector   = Filters.elemMatch("messages", Filters.eq("id", messageId.value))
+    val projection = MovementWithoutMessages.projection
+
+    val aggregates = Seq(
+      Aggregates.filter(selector),
+      Aggregates.sort(Sorts.descending("updated")),
+      Aggregates.project(projection)
+    )
+
+    mongoRetry(Try(collection.aggregate[MovementWithoutMessages](aggregates.toList)) match {
+      case Success(obs) =>
+        obs.headOption().map {
+          case None =>
+            Right((): Unit)
+          case Some(_) =>
+            Left(DuplicateMessageIdError(s"MessageId ${messageId.value} has previously been used and cannot be reused", messageId))
+        }
+      case Failure(NonFatal(ex)) =>
+        Future.successful(Left(UnexpectedError(Some(ex))))
+    })
+  }
+
+  def verifyUniqueMessageIds(messageIds: Seq[MessageId]): EitherT[Future, MongoError, Unit] = {
+    val uniqueMessageIds = messageIds.distinct
+    if (uniqueMessageIds.isEmpty) EitherT.rightT[Future, MongoError](())
+    else {
+      val uniqueMessageIdFutures = uniqueMessageIds.map(
+        messageId => isMessageIdUnique(messageId).value
+      )
+      val uniqueMessageIdResults = Future.sequence(uniqueMessageIdFutures)
```

Comment:
No, don't make a query per ID. Instead, collect all the IDs and use [$in](https://www.mongodb.com/docs/manual/reference/operator/query/in/) to select in one query.

Reply:
Done